protocol/client/messenger.go

```diff
@@ -72,71 +75,121 @@ func (m *Messenger) Stop() {
+	contact, err := m.db.GetOneToOneChat(message.SigPubKey)
+	if err != nil {
+		return errors.Wrap(err, "could not fetch chat from database")
 	}
-	for _, s := range m.public {
-		s.Stop()
+	if contact == nil {
+		contact = &Contact{
+			Type:      ContactPrivate,
+			State:     ContactNew,
+			Name:      pubkeyToHex(message.SigPubKey), // TODO(dshulyak) replace with 3-word funny name
+			PublicKey: message.SigPubKey,
+			Topic:     DefaultPrivateTopic(),
+		}
+
+		err := m.db.SaveContacts([]Contact{*contact})
+		if err != nil {
+			return errors.Wrap(err, "can't save a new contact")
+		}
 	}
+
+	_, err = m.db.SaveMessages(*contact, []*protocol.Message{&message})
+	if err == ErrMsgAlreadyExist {
+		log.Printf("Message already exists")
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "can't add a message")
+	}
+
+	return nil
 }
 
-// addStream creates a new Stream and adds it to the Messenger.
-// For contacts with public key, we just need to make sure
-// each possible topic has a stream. For a single topic
-// for all private conversations, the map will have a len of 1.
-// In the future, private conversations will have sharded topics,
-// which means there will be many conversation over a particular topic
-// but there will be more than one topic.
-func (m *Messenger) addStream(c Contact) error {
-	options, err := createSubscribeOptions(c)
+func (m *Messenger) handlePublicMessage(chatType protocol.ChatOptions, message protocol.Message) error {
+	contact, err := m.db.GetPublicChat(chatType.ChatName)
 	if err != nil {
-		return errors.Wrap(err, "failed to create SubscribeOptions")
+		return errors.Wrap(err, "error getting public chat")
+	} else if contact == nil {
+		return errors.Wrap(err, "no chat for this message, is that a deleted chat?")
+	}
+	_, err = m.db.SaveMessages(*contact, []*protocol.Message{&message})
+	if err == ErrMsgAlreadyExist {
+		log.Printf("Message already exists")
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "can't add a message")
 	}
 
-	switch c.Type {
-	case ContactPrivate:
-		_, exist := m.private[c.Topic]
-		if exist {
-			return nil
-		}
+	return nil
+}
 
-		stream := NewStream(
-			m.proto,
-			StreamStoreHandlerMultiplexed(m.db),
-		)
-		if err := stream.Start(context.Background(), options); err != nil {
-			return errors.Wrap(err, "can't subscribe to a stream")
-		}
+func (m *Messenger) handleMessageType(chatType protocol.ChatOptions, message protocol.Message) error {
+	// TODO: handle group chats messages
+	if chatType.OneToOne {
+		return m.handleDirectMessage(chatType, message)
+	}
+	return m.handlePublicMessage(chatType, message)
+}
 
-		m.private[c.Name] = stream
-	case ContactPublicRoom:
-		_, exist := m.public[c.Topic]
-		if exist {
-			return nil
+func (m *Messenger) handlePairInstallationMessageType(chatType protocol.ChatOptions, sm *protocol.StatusMessage, message protocol.PairInstallationMessage) error {
```

Comment:
Instead of passing `sm *protocol.StatusMessage` maybe `message protocol.PairInstallationMessage` should contain `SigPubKey` information as who signed the message? Or replace `sm *protocol.StatusMessage` with `sigPubKey *ecdsa.PublicKey`. It's a bit odd `protocol.PairInstallationMessage` is derived from `protocol.StatusMessage` but both are passed.

Reply:
> protocol.PairInstallationMessage should contain SigPubKey information as who signed the message?

I think we should probably find a generalized version, something like (for example:)
`{
  MessageMetadata: {SigPubKey, ID}
  Message: StatusMessage
}`

and then each method signature can be:

`handleX(metadata, actualTypeofMessage)`

what do you reckon?

Reply:
Isn't `StatusMessage` already such a wrapper? 

```go
type StatusMessage struct {
	Message   interface{}
	ID        []byte           `json:"-"`
	SigPubKey *ecdsa.PublicKey `json:"-"`
	Flags     Flags            `json:"-"`
}
```

So to have a common `handleX(meta, aactualMessage)` we can split it:
```go
type StatusMessageMetadata {
        ID        []byte           `json:"-"`
	SigPubKey *ecdsa.PublicKey `json:"-"`
	Flags     Flags            `json:"-"`
}

type StatusMessage struct {
        StatusMessageMetadata
	Message   interface{}
}
```

Reply:
yes, that works,
what's the signature of the various:
`handle....`
should we go for
`handlePairingMessage(metadata, PairingMessage)`
or 
`handlePairingMessage(StatusMessage)`
and duplicate the `type` cohercion inside? (I would go for the first one), what about you?

Reply:
I would go with the first one because dispatching is in one place and it would be clear what type of a message a given handler deals with. Fewer `interface{}` the better.